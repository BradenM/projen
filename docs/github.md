# GitHub

By default, many projects are initialized with `GitHub` component to enabled GitHub as the default provider for CI/CD workflows. See https://docs.github.com/en/actions for more information.

The use of GitHub (and generating corresponding files in `.github`) can be disabled by specifying `github: false` in your project options.

## GitHub API access

Several workflows generated by projen use APIs that are not available through
the permissions of [GITHUB_TOKEN]. To use these workflows, you must provide
either a Personal Access Token (PAT) or a GitHub App to provide API access.

[GITHUB_TOKEN]: https://docs.github.com/en/actions/security-guides/automatic-token-authentication

### Personal Access Token

Follow the [GitHub docs
instructions](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token)
for creating a personal access token. When creating the PAT, grant the token
`repo`, `workflow` and `write:packages` and `admin:repo_hook` permissions.

Add the token as a secret to your repo under the name `PROJEN_GITHUB_TOKEN`.

### GitHub App

Follow the [GitHub docs instructions](https://docs.github.com/en/developers/apps/building-github-apps/creating-a-github-app) for creating a GitHub App. Enable read & write permission for "Contents" and "Pull Request" scopes.

Add the App ID as a secret to your repo under the name `PROJEN_APP_ID` and the private key you generate as a secret to your repo under the name `PROJEN_APP_PRIVATE_KEY`.

Then, configure your projenrc file to use the GitHub app for API access:

```ts
const { NodeProject } = require('projen/javascript');

const project = new javascript.NodeProject({
  // ...other options
  githubOptions: {
    projenCredentials: github.GithubCredentials.fromApp({ ... }),
  },
});
```

## Automatic Merging

Automatic merging can be configured through the `autoMerge` and
`autoMergeOptions` fields on most default projects, or by directly creating an
`AutoMerge` component in your project. Currently this uses
[Mergify](https://mergify.com/), and expects it to be installed on your GitHub
repository for it to work. (Stay tuned for more ways of supporting auto
merging!)

To access the underlying `Mergify` configuration, you can use
`project.autoMerge.mergify`. See `AutoMerge` and `Mergify` in the API reference
for a full list of available options and methods.

## Workflows

TODO

See the `GitHub`, `GithubWorkflow`, and `Job` types in the [API
reference](./api/API.md) for currently available APIs.

Example code of creating a GitHub workflow:
<https://github.com/projen/projen/blob/65b4194c163f47ba4842981b0c92dbe516be787b/src/github/auto-approve.ts#L67-L105>

### Stale workflow

A "stale" workflow can be added which will automatically close issues or pull
requests on your GitHub repository after time has passed without the issue
seeing any comments or updates. You can enable as shown below:

```ts
// or PythonProject, etc.
new typescript.TypeScriptProject({
  stale: true,
  staleOptions: {
    issues: {
      closeMessage: "closing pull request",
      staleLabel: "I-AM-STALE",
      daysBeforeStale: 180,
    },
  }
})
```

Check the API reference for a list of all available options.

When enabled, by default issues with no activity with will be marked as stale
after 60 days and closed within 7 days, and pull requests with no activity will
be marked as stale after 14 days and closed within 2 days.
