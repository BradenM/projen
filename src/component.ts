import { Construct } from "constructs";

/**
 * Represents a project component.
 */
export class Component extends Construct {
  constructor(scope: Construct, id: string) {
    super(scope, id);
  }

  /**
   * Called before synthesis.
   */
  public preSynthesize() {}

  /**
   * Synthesizes files to the project output directory.
   */
  public synthesize() {}

  /**
   * Called after synthesis. Order is *not* guaranteed.
   */
  public postSynthesize() {}
}
