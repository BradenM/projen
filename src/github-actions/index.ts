export * from "./actions";
export * from "./github-action-metadata";
