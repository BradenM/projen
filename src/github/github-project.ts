import {
  AutoApprove,
  AutoApproveOptions,
  AutoMergeOptions,
  GitHub,
  GitHubOptions,
  MergifyOptions,
  Stale,
  StaleOptions,
} from ".";
import { Project, ProjectOptions } from "../project";

/**
 * Options for `GitHubProject`.
 */
export interface GitHubProjectOptions extends ProjectOptions {
  /**
   * Enable GitHub integration.
   *
   * Enabled by default for root projects. Disabled for non-root projects.
   *
   * @default true
   */
  readonly github?: boolean;

  /**
   * Options for GitHub integration
   *
   * @default - see GitHubOptions
   */
  readonly githubOptions?: GitHubOptions;

  /**
   * Whether mergify should be enabled on this repository or not.
   *
   * @default true
   * @deprecated use `githubOptions.mergify` instead
   */
  readonly mergify?: boolean;

  /**
   * Options for mergify
   *
   * @default - default options
   * @deprecated use `githubOptions.mergifyOptions` instead
   */
  readonly mergifyOptions?: MergifyOptions;

  /**
   * Enable and configure the 'auto approve' workflow.
   * @default - auto approve is disabled
   */
  readonly autoApproveOptions?: AutoApproveOptions;

  /**
   * Configure options for automatic merging on GitHub. Has no effect if
   * `github.mergify` is set to false.
   *
   * @default - see defaults in `AutoMergeOptions`
   */
  readonly autoMergeOptions?: AutoMergeOptions;

  /**
   * Auto-close stale issues and pull requests. To disable set `stale` to `false`.
   *
   * @default - see defaults in `StaleOptions`
   */
  readonly staleOptions?: StaleOptions;

  /**
   * Auto-close of stale issues and pull request. See `staleOptions` for options.
   *
   * @default true
   */
  readonly stale?: boolean;

  /**
   * The name of a secret which includes a GitHub Personal Access Token to be
   * used by projen workflows. This token needs to have the `repo`, `workflows`
   * and `packages` scope.
   *
   * @default "PROJEN_GITHUB_TOKEN"
   */
  readonly projenTokenSecret?: string;
}

/**
 * GitHub-based project.
 *
 * @deprecated This is a *temporary* class. At the moment, our base project
 * types such as `NodeProject` and `JavaProject` are derived from this, but we
 * want to be able to use these project types outside of GitHub as well. One of
 * the next steps to address this is to abstract workflows so that different
 * "engines" can be used to implement our CI/CD solutions.
 */
export class GitHubProject extends Project {
  /**
   * Access all github components.
   *
   * This will be `undefined` for subprojects.
   */
  public readonly github: GitHub | undefined;

  /**
   * Auto approve set up for this project.
   */
  public readonly autoApprove?: AutoApprove;

  constructor(options: GitHubProjectOptions) {
    super(options);

    // we only allow these global services to be used in root projects
    const github = options.github ?? (this.parent ? false : true);
    this.github = github
      ? new GitHub(this, {
          projenTokenSecret: options.projenTokenSecret,
          mergify: options.mergify,
          mergifyOptions: options.mergifyOptions,
          ...options.githubOptions,
        })
      : undefined;

    if (options.autoApproveOptions && this.github) {
      this.autoApprove = new AutoApprove(
        this.github,
        options.autoApproveOptions
      );
    }

    const stale = options.stale ?? true;
    if (stale && this.github) {
      new Stale(this.github, options.staleOptions);
    }
  }
}
