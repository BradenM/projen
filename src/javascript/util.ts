import { readFileSync } from "fs";
import { basename, dirname, extname, join, sep } from "path";
import * as semver from "semver";

export function renderBundleName(entrypoint: string) {
  const parts = join(entrypoint).split(sep);
  if (parts[0] === "src") {
    parts.shift(); // just remove 'src' if its the first element for ergonomics
  }

  const p = parts.join(sep);
  const dir = dirname(p);
  const base = basename(p, extname(p));
  return join(dir, base);
}

/**
 * Regex for AWS CodeArtifact registry
 */
export const codeArtifactRegex =
  /^https:\/\/(?<registry>(?<domain>[^\.]+)-(?<accountId>\d{12})\.d\.codeartifact\.(?<region>[^\.]+).*\.amazonaws\.com\/.*\/(?<repository>\w+)\/)/;

/**
 * gets AWS details from the Code Artifact registry URL
 * throws exception if not matching expected pattern
 * @param registryUrl Code Artifact registry URL
 * @returns object containing the (domain, accountId, region, repository)
 */
export function extractCodeArtifactDetails(registryUrl: string) {
  const match = registryUrl.match(codeArtifactRegex);
  if (match?.groups) {
    const { domain, accountId, region, repository, registry } = match.groups;
    return { domain, accountId, region, repository, registry };
  }
  throw new Error("Could not get CodeArtifact details from npm Registry");
}

export function minVersion(version: string): string | undefined {
  if (semver.validRange(version)) {
    return semver.minVersion(version)?.version;
  } else {
    return version;
  }
}

/**
 * Attempt to resolve the installed version of a given dependency.
 * @param dependencyName Name of dependency.
 * @param options Optional options passed through to `require.resolve`.
 */
export function tryResolveDependencyVersion(
  dependencyName: string,
  options?: { paths: string[] }
): string | undefined {
  const manifestId = `${dependencyName}/package.json`;
  try {
    const manifestPath = require.resolve(manifestId, options);
    // cannot just `require('dependency/package.json')` here because
    // `options.paths` may not overlap with this node proc's resolution paths.
    const manifest = JSON.parse(
      readFileSync(manifestPath, "utf-8").toString()
    ) as {
      version: string;
    };
    return manifest.version;
  } catch (e) {
    return undefined;
  }
}
