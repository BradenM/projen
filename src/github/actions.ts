import { Component } from "../component";
import { IResolvable } from "../file";
import { Project } from "../project";

/**
 * Manages GitHub actions used in GitHub workflows.
 */
export class Actions extends Component {
  private readonly actions: Record<string, ActionMetadata> = {};
  constructor(project: Project) {
    super(project);
  }

  /**
   * Add an action.
   *
   * Expects a fully qualified action, like "actions/checkout@v2". If
   * an action with the same version has already been specified, it will be
   * overridden.
   */
  public addAction(actionVersion: string): ActionMetadata {
    const [name, version] = actionVersion.split("@");
    if (!version) {
      throw new Error(
        `Action "${actionVersion}" needs to be tagged with a version, like "actions/checkout@v3".`
      );
    }

    const existing = this.actions[name];

    const action = { name, version };
    this.actions[name] = action;

    return existing;
  }

  /**
   * Retrieve a registered GitHub Action, throwing if no action with that
   * name is registered.
   */
  public getAction(actionName: string): ActionMetadata {
    const action = this.tryGetAction(actionName);
    if (!action) {
      throw new Error(
        `No action with name ${actionName} found. Add an action with addAction()`
      );
    }
    return action;
  }

  /**
   * Retrieve a registered GitHub action, returning `undefined` if no action
   * with that name is registered.
   */
  public tryGetAction(actionName: string): ActionMetadata | undefined {
    return this.actions[actionName];
  }

  /**
   * Returns a lazy value that will resolve to "actionName@version" based on the
   * version registered in Actions.
   *
   * If no version is registered for the given action name, either `fallback`
   * will be used if provided, or the function will throw.
   */
  public use(actionName: string, fallbackVersion?: string): string {
    const lazy: IResolvable = {
      toJSON: () => {
        if (fallbackVersion) {
          const version = this.tryGetAction(actionName)?.version;
          return `${actionName}@${version ?? fallbackVersion}`;
        } else {
          const version = this.getAction(actionName).version;
          return `${actionName}@${version}`;
        }
      },
    };
    return lazy as unknown as string;
  }
}

export interface ActionMetadata {
  readonly name: string;
  readonly version: string;
}
