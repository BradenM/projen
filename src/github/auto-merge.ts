import { Component } from "../component";
import { GitHub } from "./github";
import { Mergify } from "./mergify";

export interface AutoMergeOptions {
  /**
   * Number of approved code reviews.
   * @default 1
   */
  readonly approvedReviews?: number;

  /**
   * List of labels that will prevent auto-merging.
   * @default ['do-not-merge']
   */
  readonly blockingLabels?: string[];
}

/**
 * Sets up mergify to merging approved pull requests.
 *
 * `approvedReviews` specified the number of code review approvals required for
 * the PR to be merged.
 */
export class AutoMerge extends Component {
  /**
   * Mergify configuration for this project.
   */
  public readonly mergify: Mergify;

  private readonly lazyConditions = new Array<IAddConditionsLater>();

  constructor(github: GitHub, options: AutoMergeOptions = {}) {
    super(github.project);

    this.mergify = new Mergify(github);

    const blockingLabels = options.blockingLabels ?? ["do-not-merge"];
    const blockingCondition = blockingLabels?.length
      ? [`-label~=(${blockingLabels.join("|")})`]
      : [];

    const mergeAction = {
      delete_head_branch: {},

      queue: {
        // squash all commits into a single commit when merging
        // method: "squash",
        method: "squash",
        name: "default",
        // use PR title+body as the commit message
        commit_message_template: [
          "{{ title }} (#{{ number }})",
          "",
          "{{ body }}",
        ].join("\n"),
      },
    };

    const approvedReviews = options.approvedReviews ?? 1;

    // add initial conditions (additional conditions can be added later)
    this.addConditions(`#approved-reviews-by>=${approvedReviews}`);
    this.addConditions(...blockingCondition);

    this.mergify.addRule({
      name: "Automatic merge on approval and successful build",
      actions: mergeAction,
      conditions: (() => this.renderConditions()) as any,
    });

    this.mergify.addQueue({
      name: "default",
      conditions: (() => this.renderConditions()) as any,
    });
  }

  /**
   * Adds conditions to the auto merge rule.
   * @param conditions The conditions to add (mergify syntax)
   */
  public addConditions(...conditions: string[]) {
    this.addConditionsLater({ render: () => conditions });
  }

  /**
   * Adds conditions that will be rendered only during synthesis.
   * @param later The later
   */
  public addConditionsLater(later: IAddConditionsLater) {
    this.lazyConditions.push(later);
  }

  private renderConditions() {
    const output = new Array<string>();

    for (const later of this.lazyConditions) {
      output.push(...later.render());
    }

    return output;
  }
}

export interface IAddConditionsLater {
  render(): string[];
}
