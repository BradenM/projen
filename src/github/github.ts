import { Construct } from "constructs";
import { Component } from "../component";
import { Project } from "../project";
import { Dependabot, DependabotOptions } from "./dependabot";
import { Mergify, MergifyOptions } from "./mergify";
import { PullRequestTemplate } from "./pr-template";
import { PullRequestLint, PullRequestLintOptions } from "./pull-request-lint";

export interface GitHubOptions {
  /**
   * Whether mergify should be enabled on this repository or not.
   *
   * @default true
   */
  readonly mergify?: boolean;

  /**
   * Options for Mergify.
   *
   * @default - default options
   */
  readonly mergifyOptions?: MergifyOptions;

  /**
   * Enables GitHub workflows. If this is set to `false`, workflows will not be created.
   *
   * @default true
   */
  readonly workflows?: boolean;

  /**
   * Add a workflow that performs basic checks for pull requests, like
   * validating that PRs follow Conventional Commits.
   *
   * @default true
   */
  readonly pullRequestLint?: boolean;

  /**
   * Options for configuring a pull request linter.
   *
   * @default - see defaults in `PullRequestLintOptions`
   */
  readonly pullRequestLintOptions?: PullRequestLintOptions;

  /**
   * The name of a secret which includes a GitHub Personal Access Token to be
   * used by projen workflows. This token needs to have the `repo`, `workflows`
   * and `packages` scope.
   *
   * @default "PROJEN_GITHUB_TOKEN"
   */
  readonly projenTokenSecret?: string;
}

export class GitHub extends Component {
  /**
   * Returns the `GitHub` component of a project or `undefined` if the project
   * does not have a GitHub component.
   */
  public static of(project: Project): GitHub | undefined {
    const isGitHub = (c: Component): c is GitHub => c instanceof GitHub;
    return project.components.find(isGitHub);
  }

  /**
   * The `Mergify` configured on this repository. This is `undefined` if Mergify
   * was not enabled when creating the repository.
   */
  public readonly mergify?: Mergify;

  /**
   * Are workflows enabled?
   */
  public readonly workflowsEnabled: boolean;

  /**
   * The name of a secret with a GitHub Personal Access Token to be used by
   * projen workflows.
   */
  public readonly projenTokenSecret: string;

  public constructor(scope: Construct, options: GitHubOptions = {}) {
    super(scope, "GitHub");

    this.workflowsEnabled = options.workflows ?? true;
    this.projenTokenSecret = options.projenTokenSecret ?? "PROJEN_GITHUB_TOKEN";

    if (options.mergify ?? true) {
      this.mergify = new Mergify(this, options.mergifyOptions);
    }

    if (options.pullRequestLint ?? true) {
      new PullRequestLint(this, options.pullRequestLintOptions);
    }
  }

  public addPullRequestTemplate(...content: string[]) {
    return new PullRequestTemplate(this, { lines: content });
  }

  public addDependabot(options?: DependabotOptions) {
    return new Dependabot(this, options);
  }
}
