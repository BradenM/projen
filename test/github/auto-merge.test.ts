import { AutoMerge } from "../../src/github";
import { NodeProject, NodeProjectOptions } from "../../src/javascript";
import { synthSnapshot } from "../util";

describe("auto-merge", () => {
  test("default", () => {
    const project = createProject();

    new AutoMerge(project.github!);

    const snapshot = synthSnapshot(project);

    expect(snapshot[".mergify.yml"]).toBeDefined();
    expect(snapshot[".mergify.yml"]).toMatchSnapshot();
  });

  test("configure options", () => {
    const project = createProject();

    new AutoMerge(project.github!, {
      approvedReviews: 2,
      blockingLabels: ["draft"],
    });

    const snapshot = synthSnapshot(project);

    expect(snapshot[".mergify.yml"]).toBeDefined();
    expect(snapshot[".mergify.yml"]).toMatchSnapshot();
  });

  test("provides access to mergify", () => {
    const project = createProject();

    const autoMerge = new AutoMerge(project.github!, {
      approvedReviews: 2,
      blockingLabels: ["draft"],
    });
    autoMerge.mergify.addRule({
      name: "automatic merge for main when CI passes and 2 reviews",
      conditions: [
        "#approved-reviews-by>=2",
        "check-success=Travis CI - Pull Request",
        "base=main",
      ],
      actions: { merge: { method: "merge " } },
    });

    const snapshot = synthSnapshot(project);

    expect(snapshot[".mergify.yml"]).toBeDefined();
    expect(snapshot[".mergify.yml"]).toMatchSnapshot();
  });
});

type ProjectOptions = Omit<
  NodeProjectOptions,
  "outdir" | "defaultReleaseBranch" | "name"
>;
function createProject(options: ProjectOptions = {}): NodeProject {
  return new NodeProject({
    defaultReleaseBranch: "main",
    name: "node-project",
    autoMerge: false,
    ...options,
  });
}
