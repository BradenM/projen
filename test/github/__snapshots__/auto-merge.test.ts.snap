// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auto-merge configure options 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

queue_rules:
  - name: default
    conditions:
      - \\"#approved-reviews-by>=2\\"
      - -label~=(draft)
pull_request_rules:
  - name: Automatic merge on approval and successful build
    actions:
      delete_head_branch: {}
      queue:
        method: squash
        name: default
        commit_message_template: |-
          {{ title }} (#{{ number }})

          {{ body }}
    conditions:
      - \\"#approved-reviews-by>=2\\"
      - -label~=(draft)
"
`;

exports[`auto-merge default 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

queue_rules:
  - name: default
    conditions:
      - \\"#approved-reviews-by>=1\\"
      - -label~=(do-not-merge)
pull_request_rules:
  - name: Automatic merge on approval and successful build
    actions:
      delete_head_branch: {}
      queue:
        method: squash
        name: default
        commit_message_template: |-
          {{ title }} (#{{ number }})

          {{ body }}
    conditions:
      - \\"#approved-reviews-by>=1\\"
      - -label~=(do-not-merge)
"
`;

exports[`auto-merge provides access to mergify 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

queue_rules:
  - name: default
    conditions:
      - \\"#approved-reviews-by>=2\\"
      - -label~=(draft)
pull_request_rules:
  - name: Automatic merge on approval and successful build
    actions:
      delete_head_branch: {}
      queue:
        method: squash
        name: default
        commit_message_template: |-
          {{ title }} (#{{ number }})

          {{ body }}
    conditions:
      - \\"#approved-reviews-by>=2\\"
      - -label~=(draft)
  - name: automatic merge for main when CI passes and 2 reviews
    conditions:
      - \\"#approved-reviews-by>=2\\"
      - check-success=Travis CI - Pull Request
      - base=main
    actions:
      merge:
        method: \\"merge \\"
"
`;
